### PR TITLE
Imperative backward hooks

### PR DESCRIPTION
**Pull request by velconia**

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_imperative_basic.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+            inputs = []
+            for _ in range(10):
+                inputs.append(fluid.imperative.base.to_variable(x))
+            ret = fluid.layers.sums(inputs)
+            loss = fluid.layers.reduce_sum(ret)
+            loss._backward()
+            self.assertTrue(np.allclose(ret._numpy(), x * 10))
+            self.assertTrue(np.allclose(inputs[0]._gradient(), x))
+
+    #  def test_layer(self):
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_imperative_basic.py]: ??

File: paddle/fluid/framework/block_desc.cc [hunk 2]
```diff
@@ -155,6 +155,15 @@ void BlockDesc::RemoveOp(size_t s, size_t e) {
   ops_.erase(ops_.begin() + s, ops_.begin() + e);
 }
 
+void BlockDesc::RemoveOpInternal(const OpDesc *op_desc) {
+  for (auto it = ops_.begin(); it != ops_.end(); ++it) {
```
Comment (panyx0718) [inline paddle/fluid/framework/block_desc.cc:164]: can this be very slow?

File: python/paddle/fluid/imperative/tracer.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        # record op's trace id
+        op.iop._trace_id = self._trace_id
+
+        # trace op and save it
+        backward_refs = self.trace(op.iop, op.inputs, op.outputs, op.block.desc,
+                                   framework._current_expected_place(),
+                                   stop_gradient)
+
+        if not stop_gradient:
+            self._trace_id += 1
```
Comment (panyx0718) [inline python/paddle/fluid/imperative/tracer.py:51]: some iop can have duplicated trace_id?

File: python/paddle/fluid/initializer.py [hunk 4]
```diff
@@ -165,7 +166,8 @@ def __call__(self, var, block):
                 'force_cpu': self._force_cpu or force_init_on_cpu()
             },
             stop_gradient=True)
-        var.op = op
+        if not base.enabled():
```
Comment (panyx0718) [inline python/paddle/fluid/initializer.py]: call `base` `imperative_base` for easier read?

File: python/paddle/fluid/initializer.py [hunk 5]
```diff
@@ -165,7 +166,8 @@ def __call__(self, var, block):
...
+            var.op = op
```
Comment (panyx0718) [inline python/paddle/fluid/initializer.py:170]: why imperative mode doesn't allow it?

Comment (velconia) [reply python/paddle/fluid/initializer.py:170]: because var will refer to op and op will not be removed forever.

File: python/paddle/fluid/imperative/tracer.py (hunk 3 above)
Comment (velconia) [reply python/paddle/fluid/imperative/tracer.py:51]: iop is designed to all have -1 trace_id if they have no gradient, however, iop with grad will always have different trace_id

File: python/paddle/fluid/tests/unittests/test_imperative_basic.py (hunk 1 above)
Comment (velconia) [reply python/paddle/fluid/tests/unittests/test_imperative_basic.py]: PyLayer could be not mixed with Layer now, I will add a TODO to this

File: paddle/fluid/framework/block_desc.cc (hunk 2 above)
Comment (velconia) [reply paddle/fluid/framework/block_desc.cc:164]: I will leave a TODO here

File: python/paddle/fluid/initializer.py (hunk 5 above)
Comment (panyx0718) [reply python/paddle/fluid/initializer.py:170]: why? I thought var and op will be deleted together after backward

**Review by panyx0718 — DISMISSED** [review]

File: python/paddle/fluid/initializer.py (hunk 5 above)
Comment (velconia) [reply python/paddle/fluid/initializer.py:170]: Weight will keep in block and not delete, so the initializer will not be removed too.

File: python/paddle/fluid/initializer.py [hunk 6]
```diff
@@ -165,7 +166,8 @@ def __call__(self, var, block):
                 'force_cpu': self._force_cpu or force_init_on_cpu()
             },
             stop_gradient=True)
-        var.op = op
+        if not imperative_base.enabled():
```
Comment (JiabinYang) [inline python/paddle/fluid/initializer.py:169]: maybe use `__imperative_mode__()` to avoid potential cycle dependency

File: python/paddle/fluid/tests/unittests/test_imperative_ptb_rnn.py [hunk 7]
```diff
@@ -302,7 +304,7 @@ def test_ptb_rnn_cpu_float32(self):
             static_loss_value = None
             static_last_cell_value = None
             static_last_hidden_value = None
-            for i in range(2):
+            for i in range(batch_num):
```
Comment (JiabinYang) [inline python/paddle/fluid/tests/unittests/test_imperative_ptb_rnn.py:307]: can this train to the end?

Comment (velconia) [reply python/paddle/fluid/tests/unittests/test_imperative_ptb_rnn.py:307]: yes

File: python/paddle/fluid/initializer.py (hunk 6 above)
Comment (velconia) [reply python/paddle/fluid/initializer.py:169]: may be fixed in next pr

**Review by panyx0718 — APPROVED** [review]
